### PR TITLE
Updated models to match actual data structures transmitted by GitHub.

### LINKED ISSUE #1789: What should I name the Commit class of a PushEventPayload?
The PushEvent payload have a different `Commit` class structure than just about anywhere else (the JSON samples are from the official docs):

**PushEvent Payload:**
```json
"commits": [
    {
      "id": "0d1a26e67d8f5eaf1f6ba5c57fc3c7d91ac0fd1c",
      "tree_id": "f9d2a07e9488b91af2641b26b9407fe22a451433",
      "distinct": true,
      "message": "Update README.md",
      "timestamp": "2015-05-05T19:40:15-04:00",
      "url": "https://github.com/baxterthehacker/public-repo/commit/0d1a26e67d8f5eaf1f6ba5c57fc3c7d91ac0fd1c",
      "author": {
        "name": "baxterthehacker",
        "email": "baxterthehacker@users.noreply.github.com",
        "username": "baxterthehacker"
      },
      "committer": {
        "name": "baxterthehacker",
        "email": "baxterthehacker@users.noreply.github.com",
        "username": "baxterthehacker"
      },
      "added": [

      ],
      "removed": [

      ],
      "modified": [
        "README.md"
      ]
    }
  ],
  "head_commit": {
    "id": "0d1a26e67d8f5eaf1f6ba5c57fc3c7d91ac0fd1c",
    "tree_id": "f9d2a07e9488b91af2641b26b9407fe22a451433",
    "distinct": true,
    "message": "Update README.md",
    "timestamp": "2015-05-05T19:40:15-04:00",
    "url": "https://github.com/baxterthehacker/public-repo/commit/0d1a26e67d8f5eaf1f6ba5c57fc3c7d91ac0fd1c",
    "author": {
      "name": "baxterthehacker",
      "email": "baxterthehacker@users.noreply.github.com",
      "username": "baxterthehacker"
    },
    "committer": {
      "name": "baxterthehacker",
      "email": "baxterthehacker@users.noreply.github.com",
      "username": "baxterthehacker"
    },
    "added": [

    ],
    "removed": [

    ],
    "modified": [
      "README.md"
    ]
  },
```

**Just about anywhere else:**
```json
 {
    "sha": "47045ee9724ab11d0249bae75a599dd3ebd10a09",
    "commit": {
      "author": {
        "name": "Itai Bar-Haim",
        "email": "itaibh@gmail.com",
        "date": "2018-01-01T09:14:48Z"
      },
      "committer": {
        "name": "GitHub",
        "email": "noreply@github.com",
        "date": "2018-01-01T09:14:48Z"
      },
      "message": "Merge pull request #1 from itaibh/itaibh-patch-1\n\nUpdate README.md",
      "tree": {
        "sha": "d23387f363f3781160241279b50d03a08a6624ac",
        "url": "https://api.github.com/repos/itaibh/Spoon-Knife/git/trees/d23387f363f3781160241279b50d03a08a6624ac"
      },
      "url": "https://api.github.com/repos/itaibh/Spoon-Knife/git/commits/47045ee9724ab11d0249bae75a599dd3ebd10a09",
      "comment_count": 0,
      "verification": {
        "verified": true,
        "reason": "valid",
        "signature": "-----BEGIN PGP SIGNATURE-----\n\nwsBcBAABCAAQBQJaSfwICRBK7hj4Ov3rIwAAdHIIAIi/QCdPQ/nHCLApqXbmbbUi\nMrpd4GSHTH1BnpIZBZVfFV38/wbsEGJk+UG0YGSiX6pEWwOTapCFlCyFd5MNDtwu\nHILfCa1jgv28bwh6Nl18b7Gqx2fKq5JJeekHIO4FJyShhfTQD/nCo202cJZdwgHc\nefVwwDvJ8kTPzb1KTSDAeoKQVVHqXPSTWzTQGYF5PuqktjLhSg3fIx8UhnT+dJQF\nlqxy2GXFqax0gndC+LnjV+d13MymYY9/G82ji/CLFRzWHNbyICsnZC/7BIHzi5Iu\nKGFP0tFE3d35RFqhc5/0qw/npDjfvt0xD25ixj5i/7cumgY+LH8bU8Y0E3P2qtg=\n=GPAE\n-----END PGP SIGNATURE-----\n",
        "payload": "tree d23387f363f3781160241279b50d03a08a6624ac\nparent 8340ef42a8fc9915563396820b7cedc7424e3a84\nparent 1659dd0636941523600be861fd6015f7d48a4e76\nauthor Itai Bar-Haim <itaibh@gmail.com> 1514798088 +0200\ncommitter GitHub <noreply@github.com> 1514798088 +0200\n\nMerge pull request #1 from itaibh/itaibh-patch-1\n\nUpdate README.md"
      }
    },
    "url": "https://api.github.com/repos/itaibh/Spoon-Knife/commits/47045ee9724ab11d0249bae75a599dd3ebd10a09",
    "html_url": "https://github.com/itaibh/Spoon-Knife/commit/47045ee9724ab11d0249bae75a599dd3ebd10a09",
    "comments_url": "https://api.github.com/repos/itaibh/Spoon-Knife/commits/47045ee9724ab11d0249bae75a599dd3ebd10a09/comments",
    "author": {
      "login": "itaibh",
      "id": 140983,
      "avatar_url": "https://avatars3.githubusercontent.com/u/140983?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/itaibh",
      "html_url": "https://github.com/itaibh",
      "followers_url": "https://api.github.com/users/itaibh/followers",
      "following_url": "https://api.github.com/users/itaibh/following{/other_user}",
      "gists_url": "https://api.github.com/users/itaibh/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/itaibh/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/itaibh/subscriptions",
      "organizations_url": "https://api.github.com/users/itaibh/orgs",
      "repos_url": "https://api.github.com/users/itaibh/repos",
      "events_url": "https://api.github.com/users/itaibh/events{/privacy}",
      "received_events_url": "https://api.github.com/users/itaibh/received_events",
      "type": "User",
      "site_admin": false
    },
    "committer": {
      "login": "web-flow",
      "id": 19864447,
      "avatar_url": "https://avatars3.githubusercontent.com/u/19864447?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/web-flow",
      "html_url": "https://github.com/web-flow",
      "followers_url": "https://api.github.com/users/web-flow/followers",
      "following_url": "https://api.github.com/users/web-flow/following{/other_user}",
      "gists_url": "https://api.github.com/users/web-flow/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/web-flow/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/web-flow/subscriptions",
      "organizations_url": "https://api.github.com/users/web-flow/orgs",
      "repos_url": "https://api.github.com/users/web-flow/repos",
      "events_url": "https://api.github.com/users/web-flow/events{/privacy}",
      "received_events_url": "https://api.github.com/users/web-flow/received_events",
      "type": "User",
      "site_admin": false
    },
    "parents": [
      {
        "sha": "8340ef42a8fc9915563396820b7cedc7424e3a84",
        "url": "https://api.github.com/repos/itaibh/Spoon-Knife/commits/8340ef42a8fc9915563396820b7cedc7424e3a84",
        "html_url": "https://github.com/itaibh/Spoon-Knife/commit/8340ef42a8fc9915563396820b7cedc7424e3a84"
      },
      {
        "sha": "1659dd0636941523600be861fd6015f7d48a4e76",
        "url": "https://api.github.com/repos/itaibh/Spoon-Knife/commits/1659dd0636941523600be861fd6015f7d48a4e76",
        "html_url": "https://github.com/itaibh/Spoon-Knife/commit/1659dd0636941523600be861fd6015f7d48a4e76"
      }
    ]
  },
```

As you can see, the PushEvent payload's commit structure contains `id` instead of `sha`, and have `tree_id`, `added`, `removed` and `modified` fields.

According to here https://platform.github.community/t/payload-documentation-doesnt-match-actual-payloads/5388/2 , it seems that GitHub isn't going to fix that anytime soon, so I must create a new class, and I figured I should consult the name of it.

For the meanwhile I'll just use `Commit2` (I think I saw this structure in another payload too, so `PushCommit` is not a good name), but I'll change it once we agree on a new name.

### PR DESCRIPTION
**Pull request by itaibh**
Closes #1789

### CODE REVIEW

**Comment by ryangribble** [comment]
Unfortunately these changes will break the `PushEventPayload` class when used with the Events API

As mentioned in your linked issue, the problem is that the GitHub push webhook uses vastly different payload/objects to the push event in the events API.

What we need to do here is make "new" response models for the webhook side of things, rather than modifying the event side of things.

I'd suggest renaming all the "new" ones as `PushWebhookXXX`

So in other words
- Instead of changing the `PushEventPayload` class to have extra/conflicting fields, create a new `PushWebhookPayload` class
- Rename `CommitPayload` to `PushWebhookCommit`
- Instead of changing the `Commiter` class to have extra/conflicting fields, create a new `PushWebhookCommitter` class

**Comment by itaibh** [comment]
I just tried that, and I don't know when to create a new `PushWebhookPayload` instance.
I mean, I do need to deserialize it from a JSON coming from the server, but the name the server sends is the same as of a regular push event.
Am I missing something?

**Comment by itaibh** [comment]
@ryangribble Any news regarding my question?

**Comment by ryangribble** [comment]
im not really sure what you mean sorry.  The `push` event that we already have, is from the events API activity stream calls.  Whereas the push event you want is from a webhook.  We currently dont have any code to directly handle a webhook, however if you create the response class you could still use our deserializer to pass the json payload to, and populate the response models.  As a first step can you create the response models and push them up?  Then we can see what the best way is to deserialize the webhook payload

**Comment by itaibh** [comment]
@ryangribble I merged with latest changes on `master` with my changes. Hopefully I did it right.

All the new `PushWebhook*` classes are here, but now I need to deserialize them, and I don't know how to distinguish between the `PushWebhookEvent` payload and the `PushEvent` payload in `SimpleJsonSerializer.cs` in the `GetPayloadType(string activityType)` method.

Can you review it please and try to help me figure it out?

File: Octokit.Reactive/Clients/IObservableApplicationClient.cs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+﻿using System;
```
Comment (ryangribble) [inline Octokit.Reactive/Clients/IObservableApplicationClient.cs]: this entire file is spurious and should be removed

File: Octokit.Tests.Integration/Clients/EventsClientTests.cs [hunk 2]
```diff
@@ -333,7 +333,7 @@ public void PushEventPayloadDeserializesCorrectly()
                 Assert.NotNull(pushPayload);
                 Assert.NotNull(pushPayload.Commits);
                 Assert.Equal(1, pushPayload.Commits.Count);
-                Assert.Equal("3cdcba0ccbea0e6d13ae94249fbb294d71648321", pushPayload.Commits.FirstOrDefault().Sha);
+                //Assert.Equal("3cdcba0ccbea0e6d13ae94249fbb294d71648321", pushPayload.Commits.FirstOrDefault().Id);
```
Comment (ryangribble) [inline Octokit.Tests.Integration/Clients/EventsClientTests.cs]: this test shouldnt need to be commented out

File: Octokit/Helpers/ApiUrls.cs [hunk 3]
```diff
@@ -281,6 +281,11 @@ public static Uri NotificationSubscription(int id)
             return "notifications/threads/{0}/subscription".FormatUri(id);
         }
 
+        public static Uri AccessTokens(int installationId)
```
Comment (ryangribble) [inline Octokit/Helpers/ApiUrls.cs]: this function is unrelated to this PR and should be removed

File: Octokit/Models/Common/Committer.cs [hunk 4]
```diff
@@ -43,7 +43,7 @@ public Committer(string nodeId, string name, string email, DateTimeOffset date)
             Email = email;
             Date = date;
         }
-
+
```
Comment (ryangribble) [inline Octokit/Models/Common/Committer.cs]: no changes to this file should be necessary, so revert this whitespace change

File: Octokit/Models/Common/Committer.cs [hunk 5]
```diff
@@ -73,7 +73,7 @@ public Committer(string nodeId, string name, string email, DateTimeOffset date)
         /// </value>
         public DateTimeOffset Date { get; protected set; }
 
-        internal string DebuggerDisplay
+        internal virtual string DebuggerDisplay
```
Comment (ryangribble) [inline Octokit/Models/Common/Committer.cs]: no changes to this file should be necessary, so revert this change

File: Octokit/Models/Common/PushWebhookCommitter.cs [hunk 6]
```diff
@@ -0,0 +1,62 @@
+﻿using System;
```
Comment (ryangribble) [inline Octokit/Models/Common/PushWebhookCommitter.cs]: I would say this should live under Models/Response/ActivityPayloads with the others

**Comment by ryangribble** [comment]
In terms of the deserializer code, you might be getting hung up on some of the magic that is in there, for the events API endpoints (but that doesn't have any bearing on what you're trying to do with webhooks).  The magic stuff is to handle `Activity` responses, which have a `Payload` property of type `ActivtyPayload` with some special deserializer code that actually deserializes these as the derived classes.  

For a basic webhook, you won't use any of that stuff though... you would have your own API/webserver which receives the webhook.  You would check the `X-GitHub-Event` http header, and based on the value you would know what webhook type it is and thus what webhook payload type to use.  Then you would call the octokit deserializer, and provide the actual class you want to deserialize into.  

Eg

``` csharp
var eventHeader = "push";
var payload = "<json payload here>";

if (eventHeader == "push")
{
  var push = new SimpleJsonSerializer().Deserialize<PushWebhookPayload>(payload);
}
else if (eventHeader == "check_run")
{
  var checkRun = new SimpleJsonSerializer().Deserialize<CheckRunEventPayload>(payload);
}
```

This is the bare bones way to do it... I'm keen to think about how we could add some helpers to the library to make it easier.  The problem is it's not possible to at runtime dynamically deserialize the payload into a specific response type, that the user can know at compile time.  We can do the ugly trick like we do with the event APIs, where the user only gets a base class, which the deserializer has populated with an inheritted/derived class, but the user has to somehow know this, and cast it to the correct type, but I really dont like that.  

Another way I was thinking we could do it, is by allowing the user to register their delegate code for the given types they care about, and we could then run that.  This has the advantage of being strongly typed at compile time.  It might look something like this:

```
// Setup code
Octokit.Webhooks.RegisterWebhookHandler(WebhookType.Push, x => {
  // x is a PushWebhookPayload
});
Octokit.Webhooks.RegisterWebhookHandler(WebhookType.CheckRun, x => {
  // x is a CheckRunEventPayload
});

// When receiving a webhook
var headers = request.Headers;
var payload = request.Body;

Octokit.Webhooks.Process(header, payload); // this would call the registered event, passing the deserialized/strongly typed response model into the delegate provided by the user
```

Now whether this stuff would belong in this core library or in an addon package `Octokit.Webhooks` Im still on the fence about.  I'd actually love to provide some common middle ware type approach that could actually handle the webserver/listening part of things too, so that makes me lean towards a separate library to do this stuff

But in terms of this PR I think we can just get the response models defined, and see if you can confirm that the first code sample I provided (using the Octokit deserializer directly) is working for you.  We also have [heaps more webhook payloads to implement](https://developer.github.com/v3/activity/events/types) if you want to add any others while you're here :+1:

**Comment by itaibh** [comment]
@GMouron fine by me 😃

**Comment by GMouron** [comment]
Actually, I see that @itaibh did make the modifications asked by the code review

**Comment by ryangribble** [comment]
Hmmm I must have missed that last commit that says it addresses the review comments, I will take a look at this review again

**Comment by GMouron** [comment]
Hello, any idea of when this could land on master ?
And another slightly related question, when do you usually make a new release from master ?

Thanks :)

File: Octokit/Models/Response/ActivityPayloads/PushWebhookCommitter.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+using System.Globalization;
+
+namespace Octokit
+{
+    /// <summary>
+    /// Represents the author or committer to a Git commit. This is the information stored in Git and should not be
+    /// confused with GitHub account information.
+    /// </summary>
+    [DebuggerDisplay("{DebuggerDisplay,nq}")]
+    public class PushWebhookCommitter : Committer
```
Comment (ryangribble) [inline Octokit/Models/Response/ActivityPayloads/PushWebhookCommitter.cs]: The fields in the `PushWebhookCommitter` are not a superset of `Committer` so we shouldn't use inheritance here.  We should just make this response model contain the fields that the webhook payload will have:

``` json
      "committer": {
        "name": "baxterthehacker",
        "email": "baxterthehacker@users.noreply.github.com",
        "username": "baxterthehacker"
      },
```

Comment (shiftkey) [reply Octokit/Models/Response/ActivityPayloads/PushWebhookCommitter.cs]: Resolved in 4905d42e85b9b91c0712c5a0b5afe29757975aa7

**Comment by shiftkey** [comment]
This felt very close to landing, so I rebased the branch on `master` and addressed the last bit of feedback.

**Comment by shiftkey** [comment]
@itaibh thanks for the contribution, and for working through the review with us!

**Comment by GMouron** [comment]
Great, thank you @shiftkey. I will now be able to remove my home-built version 😉

**Comment by shiftkey** [comment]
release_notes: Added new types for webhook commit payload
